Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h

```diff
@@ -210,6 +210,34 @@ static inline void emitCatchPrologueShared(B3::Air::Code& code, CCallHelpers& ji
     jit.addPtr(CCallHelpers::TrustedImm32(-code.frameSize()), GPRInfo::callFrameRegister, CCallHelpers::stackPointerRegister);
 }
 
+static inline void prepareForTailCall(CCallHelpers& jit, const B3::StackmapGenerationParams& params, const Checked<int32_t>& newStackOffsetFromFP)
+{
+    Checked<int32_t> frameSize = params.code().frameSize();
+    Checked<int32_t> stackOffset = frameSize + newStackOffsetFromFP;
+
+    RegisterAtOffsetList calleeSaves = params.code().calleeSaveRegisterAtOffsetList();
+
+    // The frame pointer points to the previous frame, we need to reason from stackPointer + frameSize.
+    calleeSaves.adjustOffsets(frameSize);
+    jit.emitRestore(calleeSaves, MacroAssembler::stackPointerRegister);
+
+    // The return PC was saved on the stack in the tail call patchpoint.
+#if CPU(X86_64)
+    stackOffset -= Checked<int32_t>(CallFrame::returnPCOffset());
+#elif CPU(ARM64) || CPU(RISCV64)
```

Comment:
On ARM64, the link register is used as the [extended offset register](https://github.com/WebKit/WebKit/blob/main/Source/JavaScriptCore/b3/B3Common.cpp#L77)
In air tail call patchpoint, if we update lr by first loading the return address in a register, then pass it as constrained tmp we get this.

```
Air    Move %x8, %lr
asm               <1184> 0x7510143140:    mov      lr, #-296 -> 0xfffffffffffffed8
asm               <1188> 0x7510143144:    add      lr, lr, fp
asm               <1192> 0x7510143148:    ldur     x8, [lr]
asm               <1196> 0x751014314c:    mov      lr, x8
Air    Move %x8, %fp
asm               <1200> 0x7510143150:    mov      lr, #-432 -> 0xfffffffffffffe50
asm               <1204> 0x7510143154:    add      lr, lr, fp
asm               <1208> 0x7510143158:    ldur     x8, [lr]
asm               <1212> 0x751014315c:    mov      fp, x8
```
So for this patch we deal with updating lr after patchpoint generation on ARM64. I am happy to know if I miss something.

Reply:
I don't think we should be allowing macro scratch register usage here (same idea as the comment above)

Reply:
@tomoliv30 So in the prologue, I think we are fine to allow scratch register usage (no B3 tmps will be live there). Here, you should take in explicit scratch registers and allocate them in the caller.